### ISSUE TITLE
[FEA] SNMG PageRank : C++/CUDA tasks

### ISSUE THREAD

**Issue by afender** [rapidsai/cugraph#104]
The existing multi GPU PageRank code is currently in this fork/branch: https://github.com/afender/cugraph/tree/multi_pagerank.

We need to consolidate the C++/CUDA part and provide a path to benchmark the PageRank kernel at this level in cuGraph. This requires the following: 

- [x]  Bring the sort kernel (kernel `1` from PRBench in cuGraph)
- [x] Provide a **C++** path to load large datasets with MPI on multiple GPUs so that we can run tests and benchmarks on large inputs.
- [x] Deploy on DGX and run on HiBench data
- [ ] Test for memory leaks

**Improvements**
- [x] C++11 support
- [x] Fix all warnings
- [x] search and delete dead code

**Post-GTC improvements**
- [ ] Fix the path for int64_t (currently crashes)
- [ ] Replace macros by templates (to deploy 32bit and 64bit at the same time)
- [ ] Add namespace
- [ ] Support empty partitions (currently, empty partitions will try to launch kernels with 0 thread blocks )
- [ ] Support triangular matrices (currently crashes)
- [ ] double check the custom memory pool (host)
- [ ] nvlink version

**Comment by BradReesWork** [comment]
If the data is coming from cuDF, why not use the sort feature found there?

**Comment by BradReesWork** [comment]
If you are using MPI, why not use the distributed reader from PRBench?   Or just always test with DASK cuDF?

**Comment by afender** [comment]
> If the data is coming from cuDF, why not use the sort feature found there?

Plan A is to use dask-cudf sort indeed. However, some questions were raised about the reliability of this feature on multi-GPU with dask-mpi.  We only have promissing experimental results that look correct currently. Since no one from dask-cudf could gave us a clear confirmation that it is not just luck, we prepare for the case where it encounter issues on larger benchmarks.

> If you are using MPI, why not use the distributed reader from PRBench? Or just always test with DASK cuDF?

This issue is exactly about providing a C++ path based on PRBench reader and sort kernels. So, in case something goes wrong with dask (like the sort), we can still benchmark cuGraph's multi-GPU Pagerank at the C++ API level.

**Comment by afender** [comment]
> What questions?
I was told by Josh that Dask doesn’t have a multi-GPU sort.

Yep, we will continue with this plan then.

**Comment by afender** [comment]
I just pushed a `sort_edge_list` in `multi_pagerank`. I did some clean up of unnecessary things. Here is what needs to be done next:
- [ ] The sort code needs to be changed so that the input edge list is on  the device and the function returns a sorted copy of the edge list on the device as well. I think that’s all that needs to change inside the sort.
- [ ] check the `TODO` at line 801 and 807 to enable the sort.
- [ ] the output of the pagerank call needs to be allocated by the C++ API (need to sync with Ishika so that the proper change happens in the python API at the same time)
- [ ] test on unsorted input
